### PR TITLE
Fix scatter destination layout compliance check

### LINKED ISSUE #4115: Scatter destination check always inserts a relayout copy
## Problem

`_enforce_scatter_destination_layout` never actually decides whether a scatter destination is compliant. It always inserts a relayout copy, even when the destination layout is already exactly what the pass wants.

## Why

The check computes the write coordinates with `indirect_sizes=None` (`enforce_indirect_access_layout.py`):

```python
write_coords = device_coordinates(target_stl, write_dep, None)
```

When `indirect_sizes` is `None`, `compute_coordinates` skips indirect symbols entirely (`views.py`, the `else: continue` branch). So no coordinate ever carries an `IndirectAccess` marker, and the loop right below — which searches those coordinates for exactly that marker — always comes up empty.

`is_compliant` starts out `False` and is only ever set inside `if indirect_stride_idxs:`. So "could not determine" silently becomes "not compliant", and the copy is inserted unconditionally. The `compliant=` debug line lives in the same skipped branch, so the diagnostic disappears exactly when you would want it.

## Reproducer

The destination's indexed dim is at device position 0 — what the pass requires — so no copy should be inserted.

```python
import torch
import torch_spyre  # noqa: F401

torch_spyre._autoload()
torch.spyre.set_device(0)
# A layout-carrying .to() skips _lazy_init, so start the runtime first.
torch.zeros(1, dtype=torch.float16).to("spyre")

from torch_spyre._C import SpyreTensorLayout, get_device_dtype, get_elem_in_stick

SLOTS, HEADS, HD, T = 8, 2, 64, 4
eps = get_elem_in_stick(torch.float16)
sticks = HD // eps

layout = SpyreTensorLayout(
    device_size=[SLOTS, HEADS, sticks, eps],
    stride_map=[HEADS * sticks * eps, sticks * eps, eps, 1],
    device_dtype=get_device_dtype(torch.float16),
)

dst = torch.zeros(SLOTS, HEADS, HD, dtype=torch.float16).to("spyre", device_layout=layout)
src = torch.randn(T, HEADS, HD, dtype=torch.float16).to("spyre")
idx = torch.arange(T, dtype=torch.int32).to("spyre")


def scatter(dst, idx, src):
    dst.index_copy_(0, idx, src)


torch.compile(scatter, dynamic=False)(dst, idx, src)
print("max abs diff:", (dst.cpu()[:T].float() - src.cpu().float()).abs().max().item())
```

Run with `SPYRE_INDUCTOR_LOG=1 SPYRE_INDUCTOR_LOG_LEVEL=DEBUG`. Observed:

```
scatter_destination_check: inserting mutation relayout copy for buf0
```

and no `compliant=` line anywhere in the log.

## Impact

Results stay correct, so this is pure overhead. Going through vLLM on granite-3.3-8b, it adds 4 restickify ops per layer over the KV cache (1.31 MB read + 1.31 MB write each), because the cache is already handed a slot-major destination layout. Removing the copies improves end-to-end decode latency by roughly 2% in `vllm bench latency` (batch 1, input 64 / output 64, 10 iters).

## Fix

Build the real subs and sizes with `_scatter_access_subs_and_sizes` — the same helper `_insert_mutation_relayout_copy` already uses — and fall back to the copy if a size cannot be resolved.

Patch (on top of `e6aa948`): https://github.com/tdoublep/torch-spyre/tree/fix/scatter-destination-check-indirect-sizes

With that applied, the reproducer above logs `indirect_device_pos=[0], expected=[0], compliant=True` and inserts no copy. `max abs diff` is `0.0` either way.

### PR DESCRIPTION
**Pull request by avery-blanchard**
Thanks for sending a pull request! Please make sure you read the [contributing guidelines](https://github.com/torch-spyre/torch-spyre/blob/main/CONTRIBUTING.md).

#### What type of PR is this?

- [x] bug
- [ ] feature
- [ ] documentation
- [ ] cleanup

#### What this PR does:

Fixes `_enforce_scatter_destination_layout` sizing indirect symbols, which caused unnecessary relayout copies. Adds` test_scatter_copy_insertion.py `to verify compliant layouts no longer trigger restickify_ ops, plus `test_scatter_layout_helpers.py` for existing helper coverage.

#### Which issue(s) this PR is related to:
Fixes #4115

### CODE REVIEW

**Review by dgrove-oss — APPROVED** [review]
Summary

Fixes unnecessary relayout copies inserted before/after scatter ops (index_copy_, tensor-indexed assignment) whose destination already has a compliant device layout. Root cause: _enforce_scatter_destination_layout called device_coordinates(target_stl, write_dep, None), passing None for indirect-access sizes even though the scatter's index symbol size is relevant at this call site — device_coordinates's own docstring says None is only valid when indirect coordinates are irrelevant to the caller. The fix resolves proper scatter_sizes against the target's own FixedTiledLayout via a new _scatter_access_subs_and_sizes helper before calling device_coordinates, and adds a new early-return for non-FixedTiledLayout targets.

Review

BLOCKERs

none

SUGGESTIONs

- [ ] test_scatter_dim2_default_layout_needs_copy and test_scatter_batched_non_leading_dim_needs_copy (tests/inductor/test_scatter_copy_insertion.py:100-114, :167-177) wrap their real assertions in bare try: ... except Exception: pass, so a compilation exception anywhere in the path would make the test silently pass without checking anything. The latter's only in-try assertion, assertIsNotNone(restickify_count), is also vacuously true (.count() never returns None) — its own docstring admits it "documents the behavior, not a hard assertion." Not blocking because the PR's core claim (compliant destinations get zero unnecessary copies) is already covered by two other tests in the same file with hard, unguarded assertEqual(restickify_count, 0, ...) checks — but these two tests currently provide no real regression protection for the "non-compliant destination still gets a copy" side of the fix. Consider asserting on the specific expected exception type instead of Exception, or dropping the try/except and asserting the real expected outcome.

QUESTIONs

- [ ] _enforce_scatter_destination_layout (torch_spyre/_inductor/enforce_indirect_access_layout.py:467-641) now unconditionally returns early when target_fixed_tiled_layout is None but a target_stl was still found via target_buf.layouts. Previously this case fell through to the (buggy) None-sizes device_coordinates call rather than being skipped outright. Is this path reachable in practice for real scatter destinations (e.g. targets with multiple candidate layouts pre-FixedTiledLayout-selection), and if so, could skipping enforcement there mean a non-compliant destination in that category no longer gets its required relayout copy inserted at all? Worth confirming this is provably safe (or intentionally deferred/tracked) rather than an accidental behavior narrowing.

Verdict

APPROVE

**Comment by tdoublep** [comment]
Thanks for this fix - can confirm it works great!
